hail/src/main/scala/is/hail/expr/types/physical/PArray.scala

```diff
@@ -7,25 +7,28 @@ import is.hail.expr.ir.EmitMethodBuilder
 import is.hail.expr.types.virtual.TArray
 import is.hail.utils._
 import scala.reflect.{ClassTag, _}
+object PArray {
+  def apply(elementType: PType, required: Boolean = false) = new PCanonicalArray(elementType, required)
+}
 
-final case class PArray(elementType: PType, override val required: Boolean = false) extends PContainer with PStreamable {
+abstract class PArray extends PContainer with PStreamable {
   lazy val virtualType: TArray = TArray(elementType.virtualType, required)
 
-  val elementByteSize: Long = UnsafeUtils.arrayElementSize(elementType)
+  override lazy val fundamentalType: PArray = {
+    if(this.isInstanceOf[PCanonicalArray]) {
+      this
+    } else {
+      new PCanonicalArray(this.elementType, this.required)
+    }
+  }
 
-  val contentsAlignment: Long = elementType.alignment.max(4)
+  def copy(elementType: PType = this.elementType, required: Boolean): PArray
 
   override def pyString(sb: StringBuilder): Unit = {
     sb.append("array<")
     elementType.pyString(sb)
     sb.append('>')
   }
-  override val fundamentalType: PArray = {
-    if (elementType == elementType.fundamentalType)
```

Comment:
Don't understand yet why fundamental type depends on the element type. Until I do, I set this in line with PSet and PDict's interpretation of fundamental types (which is: construct a PArray).

Reply:
this is simply to avoid reallocating a type if it is already fundamental.

Reply:
This logic should be copied onto the PCanonicalArray as is, I think.